acvm/src/pwg/mod.rs

```diff
@@ -85,84 +82,109 @@ pub enum OpcodeResolutionError {
-                }
-                Err(OpcodeResolutionError::OpcodeNotSolvable(_)) => {
-                    unreachable!("ICE - Result should have been converted to GateResolution")
+                    Opcode::BlackBoxFuncCall(bb_func) => {
+                        blackbox::solve(&self.backend, &mut self.witness_map, bb_func)
+                    }
+                    Opcode::Directive(directive) => {
+                        solve_directives(&mut self.witness_map, directive)
+                    }
+                    Opcode::Block(block) | Opcode::ROM(block) | Opcode::RAM(block) => {
+                        self.blocks.solve(block.id, &block.trace, &mut self.witness_map)
+                    }
+                    Opcode::Brillig(brillig) => {
+                        BrilligSolver::solve(&mut self.witness_map, brillig)
+                    }
+                };
+                match resolution {
+                    Ok(OpcodeResolution::Solved) => {
+                        stalled = false;
+                    }
+                    Ok(OpcodeResolution::InProgress) => {
+                        stalled = false;
+                        unresolved_opcodes.push(opcode.clone());
+                    }
+                    Ok(OpcodeResolution::InProgressBrillig(oracle_wait_info)) => {
+                        stalled = false;
+                        // InProgressBrillig Oracles must be externally re-solved
+                        let brillig = match opcode {
+                            Opcode::Brillig(brillig) => brillig.clone(),
+                            _ => unreachable!("Brillig resolution for non brillig opcode"),
+                        };
+                        unresolved_brillig_calls.push(UnresolvedBrilligCall {
+                            brillig,
+                            foreign_call_wait_info: oracle_wait_info,
+                        })
+                    }
+                    Ok(OpcodeResolution::Stalled(not_solvable)) => {
+                        if opcode_not_solvable.is_none() {
+                            // we keep track of the first unsolvable opcode
+                            opcode_not_solvable = Some(not_solvable);
+                        }
+                        // We push those opcodes not solvable to the back as
+                        // it could be because the opcodes are out of order, i.e. this assignment
+                        // relies on a later opcodes' results
+                        unresolved_opcodes.push(opcode.clone());
+                    }
+                    Err(OpcodeResolutionError::OpcodeNotSolvable(_)) => {
+                        unreachable!("ICE - Result should have been converted to GateResolution")
+                    }
+                    Err(err) => return Err(err),
                 }
-                Err(err) => return Err(err),
+            }
+
+            std::mem::swap(&mut self.opcodes, &mut unresolved_opcodes);
```

Comment:
I'd like to have a little bit extra scrutiny on this (with an extra test case most likely). This looks correct but we want to ensure that the internal state is updated correctly when performing an foreign call.